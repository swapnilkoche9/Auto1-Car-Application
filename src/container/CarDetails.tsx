import React, { Fragment } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Button from '../components/Button'
import * as carsActions from '../actions/cars'
import constants from '../constants/SystemConstants'
import {CarDetailsProps,Car} from '../utils/interface'

class CarDetails extends React.Component<CarDetailsProps, {}> {
  
  componentDidMount() {
    const {
      match
    } = this.props
    this.props.carsActions.getSingleCar(match.params.stockNumber)
  }

  // add particular car to favorites list
  addFavoriteCar(car:Car) {
    debugger
    this.props.carsActions.addFavoriteCar(car)
  }

  // remove particular car to favorites list
  removeCarFromFav(car:Car) {
    debugger
    this.props.carsActions.removeFavoriteCar(car)
  }

  //  Confirm if this car is already added
  existsInFavCars(car:Car, favoriteCars:Car[]) {
    debugger
    let exists = false
    favoriteCars.forEach((favCar:Car) => {
      if (favCar.stockNumber === car.stockNumber) {
        exists = true
      }
    })
    return exists
  }

  render() {

    const {
      isFetchingCars,
      car,
      favoriteCars
    } = this.props.cars

    return (
      <Fragment>
        {isFetchingCars === false ? <div className="carImageAndDescriptionContainer">
          <div className="carHeaderImage">
            <img src={car.pictureUrl} alt='car'/>
          </div>
          <div className='carView'>
            <div className='carDetails'>
              <div className='carDescriptionContainer modalName'>
                {car.modelName}
              </div>
              <div className='carDescriptionContainer carDescription'>
                {`${constants.STOCK_LABEL}
              ${car.stockNumber} -
              ${car.mileage ? `${car.mileage.number} ${car.mileage.unit.toUpperCase()} - ` : ''}
              ${car.fuelType} -
              ${car.color}`}
              </div>
              <div className='carDescriptionContainer carAvailabilityText'>
                {constants.CAR_AVAILABILITY_TEXT}
              </div>
            </div>
            <div className='saveAndRemoveCarContainer'>
              <div className='saveAndRemovetextContainer saveCarText'>
                {constants.SAVE_CAR_TEXT}
              </div>
              <div className='saveAndRemoveButton'>
                {
                  this.existsInFavCars(car, favoriteCars)
                    ? <Button text={constants.REMOVE_TEXT} handleClick={this.removeCarFromFav.bind(this, car)} />
                    : <Button text={constants.SAVE_TEXT} handleClick={this.addFavoriteCar.bind(this, car)} />
                }

              </div>
            </div>
          </div>
        </div> : <div><h1>Loading</h1></div>}
      </Fragment>
    )
  }
}

const mapStateToProps = (state:CarDetailsProps) => {
  return {
    cars: state.cars
  }
}

const mapDispatchToProps = (dispatch:any) => {
  return {
    carsActions: bindActionCreators(carsActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(CarDetails)
