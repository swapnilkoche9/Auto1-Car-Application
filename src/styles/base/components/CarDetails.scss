
.carImageAndDescriptionContainer{
  display: flex;
  flex-direction: column;
  height: 90%;
}
.carHeaderImage{
  height: 310px;
  margin-top: $xl-size;
}
.carView {
  display: flex;
  flex-direction: row;
  justify-content: center;
  height: $content-size;
  width: $content-size;
  margin:0 auto;
}
.carDetails {
  display: flex;
  flex-direction: column;
 margin-right: $xl-size;
}
.modalName{
  font-size: $btnHeight-size;
  font-weight: 700;
  margin-top: $s-size;
  color: $dark-gray
}
.carDescriptionContainer{
  text-align: left;
  margin-bottom: $xl-size;;
}
.carDescription{
  font-size: $l-size;;
  color: $dark-gray
}
.carAvailabilityText{
  font-size: $m-size;;
  color: $dark-gray
}
.saveAndRemoveCarContainer{
  margin-left: $xl-size;;
  padding: $xl-size;;
  border: 2px solid $light-gray;
  height: 98px;
  width: 94%;
  margin-top: $s-size;;
}
.saveCarText{
 text-align: left;
 font-size: $m-size;;
 color: $dark-gray
}
.saveAndRemoveButton{
text-align: right
}

