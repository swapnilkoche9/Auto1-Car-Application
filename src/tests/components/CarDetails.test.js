import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureStore from 'redux-mock-store'
import CarDetails from '../../container/CarDetails'

Enzyme.configure({ adapter: new Adapter() })
const { shallow} = Enzyme

// create any initial state needed
const initialState = {}; 
// here it is possible to pass in any middleware if needed into //configureStore
const mockStore = configureStore();
let wrapper;
let store;
beforeEach(() => {
  //creates the store with any initial state or middleware needed  
  store = mockStore(initialState)
  wrapper = shallow(<CarDetails store={store}  />)
 })

test('should render CardDetails correctly', () => {
   expect(wrapper).toMatchSnapshot()
})

test('should render CardDetails with alt data correctly', () => {
  wrapper.setProps({
    isFetchingCars: false
  })
  expect(wrapper).toMatchSnapshot()
})


