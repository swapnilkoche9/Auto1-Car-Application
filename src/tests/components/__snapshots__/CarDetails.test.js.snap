// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CardDetails correctly 1`] = `ShallowWrapper {}`;

exports[`should render CardDetails with alt data correctly 1`] = `ShallowWrapper {}`;
